feat(CI): Docker compose for all components

klaus993 opened the pull request:
> [!NOTE]
> To test you can follow instructions on
> [`docker/README.md`](https://github.com/yetanotherco/aligned_layer/blob/docker-compose/docker/README.md)

klaus993 committed:
Add Dockerfiles

klaus993 committed:
Add config files for docker

klaus993 committed:
Add docker compose manifest with all services but the batcher

klaus993 committed:
Add batcher Dockerfile and config files

klaus993 committed:
Aggregator config fixes

klaus993 committed:
Fix mit mock tokens command

klaus993 committed:
Add localstack

klaus993 committed:
Add batcher to docker compose

klaus993 committed:
Copy all config files for anvil

klaus993 committed:
Send env vars via docker compose instead of cli parameter

klaus993 committed:
Add ca-certificates package

klaus993 committed:
Add contracts to batcher

klaus993 committed:
[WIP] make batcher work

klaus993 committed:
Fix operator docker config

klaus993 committed:
Use profiles in docker compose with groups of services

klaus993 committed:
Remove sleep hack from docker commands

klaus993 committed:
Add Makefile targets to run and kill stack with docker compose

klaus993 committed:
Add docker-build Makefile target

klaus993 committed:
Add docker compose dummy services

klaus993 committed:
Make base image build in macOS

This makes later compilations use clang rather than gcc for compiling C

klaus993 committed:
Configure operator build to use GCC and CGO

klaus993 committed:
Change order of build because of dependencies

klaus993 committed:
Remove unnecessary .env files

klaus993 committed:
Make batcher and operator compile in linux/arm64

klaus993 committed:
Send GOARCH env var via command line, according to the host's arch

klaus993 committed:
Build eigenlayer-cli separately

klaus993 committed:
Build aligned cli inside batcher

klaus993 committed:
Remove commented code

klaus993 committed:
Compile aligned

klaus993 committed:
Add Makefile targets to send burst of proofs on docker

klaus993 committed:
Add docker build specific targets

klaus993 committed:
Remove unnecesary sh -c commands

klaus993 committed:
Remove old files

klaus993 committed:
Replace - for _

entropidelic committed:
Add dockerignore files

klaus993 committed:
Add docker-compose.yml to dockerignore

klaus993 committed:
Remove foundry from base image

klaus993 committed:
Batcher: run cargo install apart from build

klaus993 committed:
Batcher: remove unnecesary files

klaus993 committed:
Use env vars for RPC URL and proof generator address

klaus993 committed:
Add compilation of gnark_groth16_bn254_infinite_script in batcher

klaus993 committed:
Add Makefile targets to send different proofs with docker

klaus993 committed:
Add healtcheck to anvil and check for it on docker_up

klaus993 committed:
Remove aggregator from base profile

klaus993 committed:
Add Makefile targets to restart aggregator, operator, batcher

klaus993 committed:
Give a way to attach to foundry container to run cast

klaus993 committed:
Add basic readme for docker compose

klaus993 committed:
Add docker logs Makefile targets

klaus993 committed:
Test

klaus993 committed:
Add flags to force recreation and removing orphans

klaus993 committed:
Add Makefile targets to attach to containers

klaus993 committed:
Add Makefile target to verify proofs onchain

klaus993 committed:
Workflow test

klaus993 committed:
Add sudo

klaus993 committed:
Swap `==` for `=` for sh comparison

klaus993 committed:
Fix workflow

klaus993 committed:
Add timeout minutes 30

samoht9277 committed:
Added docker_verify_proof_submission_success make target to check proof verification success.

samoht9277 committed:
Removed halo2 proofs, fixed minor issues on docker Make targets.

klaus993 committed:
Use images from our own registry

klaus993 committed:
Remove logs from workflow

samoht9277 committed:
Added login to GitHub Container Registry

samoht9277 committed:
Chenged PAT name.

klaus993 committed:
Use docker images from registry in FROM

samoht9277 committed:
Added target to run all proof submit, also tuned the docker_verify_proof_submission_success target.

samoht9277 committed:
Merge branch 'docker-compose' of github.com:yetanotherco/aligned_layer into docker-compose

samoht9277 committed:
Added docker_batcher_send_groth16_burst, and included it to the run all proofs target.

klaus993 committed:
Add permissions to GITHUB_TOKEN built in token

klaus993 committed:
Remove sudo

klaus993 committed:
Add job post cleanup

klaus993 committed:
Send proofs batches in CI

klaus993 committed:
Verify all sent proofs in CI

klaus993 committed:
Add sleep after docker_up

klaus993 committed:
Improve naming for the new github workflow

klaus993 committed:
Sleep for 2 minutes

klaus993 committed:
Make post steps run always

klaus993 committed:
Add continue-on-error: true

klaus993 committed:
Increase wait

klaus993 committed:
Decrease wait

klaus993 committed:
Rename workflow

klaus993 committed:
Test docker build

klaus993 committed:
Parametrize repository name, image to push and branch

klaus993 committed:
Fix runs-on

klaus993 committed:
Use staging in FROM

klaus993 committed:
XXX: Build aligned_base testnet

klaus993 committed:
Swap back to ubuntu-latest

samoht9277 committed:
Added fix to docker build.

samoht9277 committed:
Added `.` to command.

samoht9277 committed:
Added `ls` and `pwd` to debug error.

klaus993 committed:
Revert "XXX: Build aligned_base testnet"

This reverts commit 74bd6ff22dc246a20c8091df611b08f5f9a6a0bd.

klaus993 committed:
Add docker build to CI

klaus993 committed:
Change tag to latest

klaus993 committed:
Fix runs-on for cleanup

klaus993 committed:
Clone with submodules outside of docker

klaus993 committed:
Remove timeout-minutes

klaus993 committed:
Add .github folder to .dockerignore

klaus993 committed:
Remove unused docker build

It will be added later in another PR

klaus993 committed:
Reduce repetitions

samoht9277 committed:
Changed docker pull repo URL where images are located.

samoht9277 committed:
Added Makefile to docker.

samoht9277 committed:
Fixed bug in COPY instruction.

samoht9277 committed:
Added batcher/aligned-sdk into docker.

samoht9277 committed:
Replaced lambdaclass with yetanotherco.

samoht9277 committed:
Added go lock files.

samoht9277 committed:
Added required dependencies on operator dockerfile.

samoht9277 committed:
Added more required dependencies on operator dockerfile.

samoht9277 committed:
Added go.mod and go.sum to batcher in docker.

klaus993 committed:
Copy batcher and go files

klaus993 committed:
Copy infinite script to batcher docker image

samoht9277 committed:
Added aggregator directory in docker.

samoht9277 committed:
Added core and metrics directory in docker.

klaus993 committed:
Add aggregator build to docker_build target

samoht9277 committed:
Added bindings to dockerfile.

samoht9277 committed:
Removed COPY from builder.

samoht9277 committed:
Added yaml.

samoht9277 committed:
Added output directory.

samoht9277 committed:
Added contracts directory.

samoht9277 committed:
Added script directory.

samoht9277 committed:
Added anvil json file.

samoht9277 committed:
Added test_files directory.

klaus993 committed:
Add pull request trigger and concurrency rules

klaus993 committed:
Merge branch 'staging' into docker-compose

IAvecilla commented on config-files/config-aggregator-docker.yaml:
```diff
@@ -0,0 +1,53 @@
+#   max_proof_size: 67108864 # 64 MiB
+#   max_batch_size: 268435456 # 256 MiB
+#   eth_ws_reconnects: 99999999999999
+#   pre_verification_is_enabled: true
```
We should remove this

klaus993 replied:
Removed!
https://github.com/yetanotherco/aligned_layer/pull/1044/commits/c52dc2cb0e8fd140dde5963c19696017aefd89d8

IAvecilla commented on config-files/config-aggregator-docker.yaml:
```diff
@@ -0,0 +1,53 @@
+# private_key_store_path: config-files/anvil.ecdsa.key.json
+# bls_private_key_store_path: config-files/anvil.bls.key.json
+# signer_type: local_keystore
+# chain_id: 31337
```
Ditto

klaus993 replied:
Removed! https://github.com/yetanotherco/aligned_layer/pull/1044/commits/c52dc2cb0e8fd140dde5963c19696017aefd89d8

klaus993 committed:
Add BATCHER_PAYMENTS_CONTRACT_ADDRESS

klaus993 committed:
Remove deprecated flag

klaus993 committed:
Remove unnecessary env var

IAvecilla commented on Makefile:
```diff
@@ -641,6 +641,221 @@ tracker_dump_db:
+docker_verify_proof_submission_success: 
+	@echo "Verifying proofs were successfully submitted..."
+	docker exec $(shell docker ps | grep batcher | awk '{print $$1}') \
+	sh -c ' \
```
I believe we should reduce the size of the Makefile by moving all this kind of logic into new bash scripts to keep it more readable. We can handle this in a separate PR since it's not very urgent.

klaus993 replied:
Maybe you're right, when I wrote this I considered that putting it into a separate shell script adds indirection and difficulty to understand what the target does, but maybe this way it's also harder to debug since it's a multi-line string and it's kinda shady.

I can create an issue and modify it later, in another PR. What do you think?

klaus993 replied:
https://github.com/yetanotherco/aligned_layer/issues/1240

IAvecilla replied:
Perfect!

Oppen commented on .github/workflows/send-proofs-docker.yml:
```diff
@@ -0,0 +1,59 @@
+        password: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Checkout
+      uses: actions/checkout@v1
```
Why are we using v1? Isn't that version abandoned?

samoht9277 replied:
True. We should use `v4` as stated [here](https://github.com/actions/checkout).

klaus993 replied:
Fixed! https://github.com/yetanotherco/aligned_layer/pull/1044/commits/7a3a2e036d737181c9dedf2d4faa839549e519af

Oppen commented on Makefile:
```diff
@@ -641,6 +641,221 @@ tracker_dump_db:
 		docker exec -t tracker-postgres-container pg_dumpall -c -U tracker_user > dump.$$(date +\%Y\%m\%d_\%H\%M\%S).sql
 	@echo "Dumped database successfully to /operator_tracker"
 
+ARCH := $(shell uname -m)
+ifeq ($(ARCH), x86_64)
+  GOARCH := amd64
+else ifeq ($(ARCH), arm64)
+  GOARCH := arm64
+endif
```
Go already defaults to the host's arch, so I'm not sure why we add this check.

klaus993 replied:
This was linked to a problem related with your comment about the host architecture for Golang installation (https://github.com/yetanotherco/aligned_layer/pull/1044#discussion_r1799927936)

Already removed: https://github.com/yetanotherco/aligned_layer/pull/1044/commits/6c43f928a3cb7b24bd8619eff9c5abea233f6bfe

Oppen commented on docker/aligned_base.Dockerfile:
```diff
@@ -0,0 +1,40 @@
+                   libssl-dev \
+                   yq \
+                   jq
+RUN wget https://golang.org/dl/go$GO_VERSION.linux-amd64.tar.gz
+RUN tar -C /usr/local -xzf go$GO_VERSION.linux-amd64.tar.gz
```
I believe this should use whatever the host arch is.

klaus993 replied:
Fixed! https://github.com/yetanotherco/aligned_layer/pull/1044/commits/4c5563b16aae8864f9dd8a917430237ce66013d8

Oppen reviewed (commented):
Looks good. I'll try the commands in the makefile later today before giving approval.

Oppen commented on docker/README.md:
````diff
@@ -0,0 +1,109 @@
+To build the necessary images, first run:
+
+```shell
+make docker_build
````
Running this command on my Mac results in:
<img width="1671" alt="image" src="https://github.com/user-attachments/assets/827b62d4-4a12-4b43-8f13-b7a264b88e81">

klaus993 replied:
Fixed! It was related to your other comment https://github.com/yetanotherco/aligned_layer/pull/1044#discussion_r1799927936

klaus993 committed:
Remove commented configurations

klaus993 committed:
Fix actions/checkout version

klaus993 committed:
Add GOARCH to aligned_base image

klaus993 committed:
Add Makefile target to build base image

klaus993 committed:
Add needed parameter

klaus993 committed:
Remove GOARCH env var usage from operator

samoht9277 committed:
Added golang support for local machine architecture.

klaus993 committed:
Remove GO env vars from batcher

klaus993 committed:
Remove GOARCH from aligned_base and from Makefile

klaus993 committed:
Remove unnecessary cargo install

Oppen commented:
The commands fill my disk, but the CI looks gooooooood!

klaus993 committed:
Add more files to .dockerignore

klaus993 committed:
Add private key to send bursts to pay for the transactions

klaus993 committed:
Revert some dockerignore changes

klaus993 committed:
Add env var to fix potential cargo problem

klaus993 committed:
Revert "Revert some dockerignore changes"

This reverts commit 96cb5b6e5041346bec3e6278e1b1561e3b31bb5f.

klaus993 committed:
Revert "Add more files to .dockerignore"

This reverts commit 00c4a2c5a5ce43de60970c2a3a8bfcb706c5fa15.

Oppen reviewed (changes requested):
I'll leave this request changes so I don't forget to branch and try to make the CI fail. A CI that doesn't fail when we make a mistake wouldn't be useful.

klaus993 committed:
Add mandatory parameter

klaus993 committed:
Merge branch 'staging', remote-tracking branch 'origin' into docker-compose

uri-99 commented on Makefile:
```diff
@@ -672,6 +672,222 @@ tracker_dump_db:
+	docker compose -f docker-compose.yaml --profile batcher build
+	docker compose -f docker-compose.yaml --profile aggregator build
+
+docker_up:
```
`make docker_up` returns the following error:

<img width="996" alt="image" src="https://github.com/user-attachments/assets/04a8e2b8-635b-444e-97e9-3218dcc68333">

I think it is fixable by increasing the gas_price of user_fund_payment_service

klaus993 replied:
Did you try again? I found this happens randomly from time to time. 

>I think it is fixable by increasing the gas_price of user_fund_payment_service

I'm interested on the fix though, does this fix it permanently, you think?

klaus993 replied:
Fixed in: https://github.com/yetanotherco/aligned_layer/pull/1044/commits/69ffbc2b2d9798963e7bc86d65451d6febc3bc38

uri-99 commented on Makefile:
```diff
@@ -672,6 +672,222 @@ tracker_dump_db:
+	@$(MAKE) docker_batcher_send_plonk_bls12_381_burst
+	@$(MAKE) docker_batcher_send_groth16_burst
+
+docker_batcher_send_infinite_groth16:
```
Any reason  why this is not a .sh script? I saw you made an issue, I think it would be best to migrate it now and avoid increasing the technical debt so much. Anyhow it should be pretty straight forward

klaus993 replied:
We don't want to delay merging this anymore. We need the CI working ASAP. I can do it with high priority if needed, but doing it in this PR would take another whole feedback loop with the review/test approval and such, and we've seen that takes some time.

klaus993 replied:
FYI https://github.com/yetanotherco/aligned_layer/issues/1240

uri-99 reviewed (changes requested):
Sorry, meant to req changes

klaus993 committed:
Fix edge case when no proofs are saved into ./aligned_verification_data

This fix makes this edge case fail the CI, as expected

klaus993 committed:
Merge branch 'staging' into docker-compose

klaus993 committed:
Add gas price to fund-operator

As it is the first cast that is run, the gas price is too expensive and
sometimes fails

klaus993 committed:
Install rust nightly

klaus993 committed:
Install stable version also

klaus993 committed:
Try with --no-self-update flag

klaus993 committed:
Use rust as base image and install nightly

klaus993 committed:
Merge remote-tracking branch 'origin/staging' into docker-compose

klaus993 committed:
Add logic to check if all proofs were sent

klaus993 committed:
Merge remote-tracking branch 'origin/staging' into docker-compose

klaus993 committed:
Remove unnecessary --batcher_url flag

As localhost is already the default

klaus993 committed:
Remove unnecessary sleep

klaus993 committed:
Decrease sleep

klaus993 committed:
Add new aggregator configs

From https://github.com/yetanotherco/aligned_layer/blob/2c11cf12bd05bce5a37a6b0a8bcfb24b6680d198/config-files/config-aggregator.yaml#L38-L40

klaus993 committed:
Add sleep

klaus993 committed:
Parametrize sleep time for proofs

klaus993 committed:
Add md files to dockerignore

klaus993 committed:
Merge remote-tracking branch 'origin/staging' into docker-compose

Oppen reviewed (approved):
Tested by the demolition crew.

klaus993 commented:
> Tested by the demolition crew.

Thanks!
BTW, how can I join that guild?

JuArce reviewed (approved):
LGTM. Just two comments 
- Update readme for send proofs targets
- Calculate gas price in every transaction sent on start up

klaus993 committed:
Add gas price to all cast commands

klaus993 committed:
Add quotes to smart contract function call

klaus993 committed:
Fix race condition for cleanup steps

As a different job, when another PR has the send proofs job queued, it's
executed before the cleanup

klaus993 committed:
Update README.md

klaus993 committed:
Add paths ignore for md files

klaus993 committed:
Fix docker_verify_proofs_onchain

klaus993 committed:
Add docker_verify_proofs_onchain to readme

klaus993 committed:
Remove confusing target from readme
